### DISCUSSION TITLE
IPv6 WSGI server fails to start

### DISCUSSION THREAD

**Discussion by ddundi0** [falconry/falcon#2187] [General]
Hi, 

I've been using falcon3.1.1 and trying to start the server on an IPv6 machine. Whatever I've tried, starting the server listening on all interfaces ('::'), or providing the actually address of the NI, it always immediately fails to start with the socket.gaierror and an "Address family for hostname not supported" exception message. 

Here's the full stack trace;
```
> Traceback (most recent call last):
  File "//program.py", line 33, in <module>
    with make_server('::', 8000, app) as httpd:
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/wsgiref/simple_server.py", line 154, in make_server
    server = server_class((host, port), handler_class)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/socketserver.py", line 456, in __init__
    self.server_bind()
  File "/usr/lib64/python3.11/wsgiref/simple_server.py", line 50, in server_bind
    HTTPServer.server_bind(self)
  File "/usr/lib64/python3.11/http/server.py", line 136, in server_bind
    socketserver.TCPServer.server_bind(self)
  File "/usr/lib64/python3.11/socketserver.py", line 472, in server_bind
    self.socket.bind(self.server_address)
socket.gaierror: [Errno -9] Address family for hostname not supported
```
It's enough to take the WSGI [Quickstart](https://falcon.readthedocs.io/en/stable/user/quickstart.html) example and provide the '::' Ipv6 address for starting the server, but anyways here's the CP of the full code.

```py
from wsgiref.simple_server import make_server

import falcon

class ThingsResource:
    def on_get(self, req, resp):
        resp.status = falcon.HTTP_200  # This is the default status
        resp.content_type = falcon.MEDIA_TEXT  # Default is JSON, so override
        resp.text = (
            '\nTwo things awe me most, the starry sky '
            'above me and the moral law within me.\n'
            '\n'
            '    ~ Immanuel Kant\n\n'
        )

app = falcon.App()

things = ThingsResource()

app.add_route('/things', things)

if __name__ == '__main__':
    with make_server('::', 8000, app) as httpd:
        print('Serving on port 8000...')
        httpd.serve_forever()`
```
Am I missing something here ? Does falcon library expose the possibility to configure address_family to something like "socket.AF_INET6". 

Also, when trying to start it on a specific address and using the python's socket API to get that address' info, the "socket.getaddrinfo" call is successful, returning something like this;
```
>  [(<AddressFamily.AF_INET6: 10>, <SocketKind.SOCK_STREAM: 1>, 6, '', ('2002:14ba:9ea:8800:9c7b:4e6c:dae2:741c', 8080, 0, 0)), (<AddressFamily.AF_INET6: 10>, <SocketKind.SOCK_DGRAM: 2>, 17, '', ('2002:14ba:9ea:8800:9c7b:4e6c:dae2:741c', 8080, 0, 0)), (<AddressFamily.AF_INET6: 10>, <SocketKind.SOCK_RAW: 3>, 0, '', ('2002:14ba:9ea:8800:9c7b:4e6c:dae2:741c', 8080, 0, 0))]
```
But starting the server on that same IP just won't work.

**Comment by CaselIT** [comment]
Hi,

The address is not controlled by falcon in any way. The same code results in the same error when using a function as the app:
```py
from wsgiref.simple_server import make_server

def app(env, start)->list[bytes]:
    start('200 ok', [])
    return [b'hi there']

if __name__ == '__main__':
    s = make_server('::', 8000, app)
    s.serve_forever()
```

I'm not sure how you can specify an ipv6 when using simple_server

**Reply by ddundi0** [reply]
I would expect nothing else but to get the same issue when using function as an app, not sure why it should make any difference.

I'm not sure If I'm following you, but what you're saying is that wsgiref.simple_server simply doesn't support this kind of startup ? That's what I suspected, I was just wondering if maybe it's configurable in a way, just that I'm missing it.

Also, I guess that invoking "make_server" with '' or 'localhost' should be working no matter the stack, but what we're having is pretty complex mesh of services and due to network routing rules we need this server to be serving on a specific NI/Address. Hardly can I see the way to work around it.

All in all, if that's the conclusion, then it is what it is. We'll look for a way to solve it. Thanks.

**Reply by CaselIT** [reply]
> I'm not sure If I'm following you, but what you're saying is that wsgiref.simple_server simply doesn't support this kind of startup ? That's what I suspected, I was just wondering if maybe it's configurable in a way, just that I'm missing it.

the point was that this seems the behaviour of simple server, not something specific to falcon.
You can enable them by setting `WSGIServer.address_family = socket.AF_INET6` before calling make server

if you use a "real" wsgi server, such as waitress or gunicorn you should be able to run the falcon app using ipv6 addresses by default

**Reply by vytas7** [reply]
Yeah, using Gunicorn I have no issues:
```
gunicorn -b [::]:8000 test:app
```
then
```
$ curl  [::1]:8000

{"title": "404 Not Found"}
```

**Comment by ddundi0** [comment]
Thanks for the advice! Anyway, as I said, it's a complex service and the backend around it, so switching to gunicorn wouldn't be that straightforward in terms of quickness to a solution. I might try to configure it as @CaselIT proposed. I'll close the discussion since I don't expect anything else here.
